### PR TITLE
Add tagging tracer and collect/inject transforms

### PR DESCRIPTION
**Pull request by jekbradbury**
## TL;DR:

This lets you name intermediate values in your computations, then extract and overwrite them using transformations. It can be used on its own—for example, to collect metrics or compute gradients of intermediate values—but also provides a JAX-native foundation for NN or PPL libraries that might compose better with other transforms or the JAX PRNG than similar functionality implemented at the Python level.

## Details:

First, we effectively augment the jaxpr language with two new (and in principle orthogonal) features:

1. A way (`lax.yield_value`) to mark that an intermediate value in a jaxpr is a yield point where evaluation can be "paused" and the value can be inspected or replaced from outside
2. A way (`lax.push_tag`) to accumulate stacks of compile-time metadata alongside values, and rules for propagating these metadata stacks through a computation's data flow

Together, these mechanisms give rise to a notion of _tagged intermediate values_ within computations: yield points whose metadata stacks each carry a name atop a stack of nested "namespaces" or "scopes". For instance, a yield point with the metadata stack ("foo", "bar") is a tagged intermediate value with the name `bar` inside a namespace `foo`.

Then, the new tagging tracer (`interpreters/tagging.py`) performs metadata stack propagation and handles yield points using their names and namespaces. When wrapped in a tagging transformation, a JAX function has an extra return value—a nested dict containing values collected at tagged intermediates—and an extra argument, a nested dict whose values will be used to replace tagged intermediates. The key paths to values in these dictionaries correspond to the names/namespaces in the metadata stacks of the tagged intermediates: for instance, the path to `1.` in `{"foo": {"bar": 1.}}` is the metadata stack `("foo", "bar")`.

The public API for the tagging tracer is a pair of function transformations, `collect` and `inject`, that respectively allow extracting and overwriting tagged intermediates. If we start with
```python
def foo(x):
  y = lax.tag(x ** 2, "y")
  z = y + 1
  return z
```
then `value, intermediates = collect(foo)(2.)` will give `intermediates = {"y": 4.}` (the value for `x**2`, the tagged intermediate named "y"), while `inject(foo, {"y": 1.})(2.)` will replace `x**2` with `1.` and return `2.`.

## Use cases:

**More convenient metrics and logging** using `collect` (avoids having to manually thread metric values out of potentially deeply nested functions)

**Gradients of intermediate values** can be obtained by injecting perturbations onto those intermediates, then differentiating with respect to the perturbations:
```python
def grad_intermediates(f):
  def grad_fun(*args, **kwargs):
    _, orig_tree = collect(f)(*args, **kwargs)
    def fun_for_grad(perturbation_tree):
      return inject(f, perturbation_tree, lambda old, new: old + new)(*args, **kwargs)
    return grad(fun_for_grad)(tree_map(np.zeros_like, orig_tree))
  return grad_fun
```
A version of this function is included as a public API.

A potential foundation for **probabilistic programming abstractions**: if sampling from distributions always takes place at a yield point, then arbitrary subsets of ancestral samples can be collected and replaced by inference algorithms.

Based on that, and along the lines of Dougal's PRNG key split tree idea (#1341), an unusual but (to me) compelling approach to **deep learning abstractions**. This involves representing a neural net architecture as a single function that initializes the network's parameters, tags them as intermediates with a namespaced structure, then uses them in a forward pass with the input data (so essentially a probabilistic program for a "Bayesian neural net").

The notion familiar from stax of a network architecture as a pair of `init` and `apply` functions can then be recovered by applying `collect` and `inject` respectively to the architecture function. Meanwhile, the namespaced structure can come from PRNG key `split` and `fold_in` operations (this is the stax2 idea, and the tagging tracer implements this behavior for PRNG key metadata stacks) or even from scopes maintained dynamically in Python (as in DeepMind's Sonnet library).

More to come on this front!

## Open questions:

The tagging tracer's use of **nested dictionaries** to represent values indexed by stacks of keys is essentially the same abstraction exposed in the `_with_path` functions in DeepMind's recently open-sourced [tree](https://github.com/deepmind/tree) library, albeit without support for lists. It might be reasonable to replace the hand-rolled helpers I added with a dependency on `tree` if we're willing to take one, or alternatively to build a notion of paths into JAX's existing pytree infrastructure.

**Composition** with semantics-preserving transformations (jit) or when nested inside semantics-non-preserving ones is well-defined (and seems to work), but there's a choice of what to do when nesting outside semantics-nonpreserving transformations like jvp and batching. For instance, should `collect(jvp(f))` only collect the primal values of named intermediates, or both the primals and tangents (and if the latter, how should they be returned)? Similarly, I think the right thing to do in inject/collect of vmap is to accept/return batched intermediate values with the batch dimension in the vmap's `out_dim`, but that might be a little surprising for `inject`. (The current behavior of this code is to collect only the primal values in a jvp trace and to collect the batching tracers themselves in batching traces).

### CODE REVIEW

**Comment by shoyer** [comment]
> there's a choice of what to do when nesting outside semantics-nonpreserving transformations like jvp and batching.

For use cases like extracting intermediate values from numerical algorithms like linear solves, it is essential to be able to get this information out of both the forwards and backwards pass.

It would be really nice to have at least an opt-in API for this.

**Comment by gehring** [comment]
>  For instance, should collect(jvp(f)) only collect the primal values of named intermediates, or both the primals and tangents (and if the latter, how should they be returned)?

I believe I would expect output to either only contain the intermediate values of tangents/co-tangents (for example, if some custom jvp implementation pushed some intermediates), or have it return both the primal's and tangent's. For the former, you could still get the primals if you wanted to if a `value_and_jvp` function was available.

With only reading the text above, here are the semantics I might expect. By default, primitives should simply propagate intermediates, i.e., identity on intermediates (which I believe is the case here), but I agree that what higher order operators should do is less clear. I think if you say that differentiation wipes all intermediates, you might get jvps returning only tangent intermediates while if you say that differentiation propagates intermediates, you would likely get jvp returning both primal and tangent. It's not quite clear what kind of low-level rule might lead to only getting intermediates of the primals.

There are probably lots of edge cases/caveat that I've missed but I thought sharing my first impressions will probably provide a better approximation of what a new user might think.

> Similarly, I think the right thing to do in inject/collect of vmap is to accept/return batched intermediate values with the batch dimension in the vmap's out_dim, but that might be a little surprising for inject. (The current behavior of this code is to collect only the primal values in a jvp trace and to collect the batching tracers themselves in batching traces).

Are you talking about when inject is used within a batched function? In that case, would batching inject be problematic?

**Comment by mgbukov** [comment]
@jekbradbury are you planning to merge these features into master eventually?

**Comment by ethanluoyc** [comment]
It would be great if this gets merged at some point. I found this to be really useful for my workflow!

**Comment by mgbukov** [comment]
> It would be great if this gets merged at some point. I found this to be really useful for my workflow!

I second that!

**Comment by mattjj** [comment]
I think the trail has gone cold on this one...
